### DISCUSSION TITLE
Is there any API to interact with the eCAL Player?

### DISCUSSION THREAD

**Discussion by piyush-m0** [eclipse-ecal/ecal#1786] [Q&A]
**Language: `C++`**

Can someone help me with interacting with the eCAL Player? 

I wish to use the `ecal_meas_cutter` tool for trimming eCAL recordings, but if I use the eCAL Player and eCAL Recorder, if the Player is paused midway recording, a wrong trimmed recording is saved since the recording will think of it as frame drops.

Is there any `C++` API to interact with the eCAL Player to take the relative timestamp of ongoing recording being played so that I can pass them to the `ecal_meas_cutter` tool as an argument as start and end timestamps?

Please let me know if Monitoring API can help me with this as well by any change.

**Comment by DownerCase** [comment]
To directly answer the title, yes eCAL player (and most of the applications) have an eCAL services API for remote control; though I didn't fully understand your problem to say if it's a good solution. Primarily confused about the relevance of eCAL player/recorder as `ecal_meas_cutter` operates directly on existing recordings.

The monitoring API could let you figure out the currently active (with synchronization time error) topics.

Intro to eCAL services:
https://eclipse-ecal.github.io/ecal/getting_started/services.html

The eCAL monitor is able to show what services an application advertises:
https://eclipse-ecal.github.io/ecal/getting_started/monitor.html#services

Samples:
https://github.com/eclipse-ecal/ecal/tree/support/v5.13/samples/cpp/services/ecalplayer_client
https://github.com/eclipse-ecal/ecal/tree/support/v5.13/samples/cpp/services/ecalplayer_gui_client

Source Protobuf defining the eCAL player services:
https://github.com/eclipse-ecal/ecal/blob/support/v5.13/app/app_pb/src/ecal/app/pb/play/service.proto

**Reply by piyush-m0** [reply]
To explain my situation in simpler words, I might be looking to build a simple gui for the ecal measurement cutter. It would be great if the Player itself would integrate but since currently it doesn't I would have to do it myself.

**Reply by DownerCase** [reply]
Ah, yeah. Sounds cool.
In that case you'd probably also want to use the measurement API so that you can access the recording file and figure out the begin and end timestamps.
https://github.com/eclipse-ecal/ecal/blob/support/v5.13/samples/cpp/measurement/measurement_read/src/measurement_read.cpp

**Reply by piyush-m0** [reply]
Will try it out, thanks!

**Reply by piyush-m0** [reply]
> Ah, yeah. Sounds cool. In that case you'd probably also want to use the measurement API so that you can access the recording file and figure out the begin and end timestamps. https://github.com/eclipse-ecal/ecal/blob/support/v5.13/samples/cpp/measurement/measurement_read/src/measurement_read.cpp

Hi,

So for getting the begin and end timestamps, I would need to use the GetMaxTimestamp and GetMinTimestamp methods of the [`eCAL::experimental::measurement::hdf5::Reader`](https://github.com/eclipse-ecal/ecal/blob/support/v5.13/contrib/measurement/hdf5/include/ecal/measurement/hdf5/reader.h) class right? This is part of the include header file `<ecal/measurement/hdf5/reader.h>`

The [sample you shared above](https://github.com/eclipse-ecal/ecal/blob/support/v5.13/samples/cpp/measurement/measurement_read/src/measurement_read.cpp) has the structure `IMeasurement` from the header file `<ecal/measurement/imeasurement.h>`. Is this one relevant?

**Reply by piyush-m0** [reply]
> So for getting the begin and end timestamps, I would need to use the GetMaxTimestamp and GetMinTimestamp methods of the [`eCAL::experimental::measurement::hdf5::Reader`](https://github.com/eclipse-ecal/ecal/blob/support/v5.13/contrib/measurement/hdf5/include/ecal/measurement/hdf5/reader.h) class right? This is part of the include header file `<ecal/measurement/hdf5/reader.h>`

I am using this include header, and I am encountering build errors:

```bash
/usr/bin/ld: CMakeFiles/measurement_cutter.dir/src/main.cpp.o: in function `main':
main.cpp:(.text.startup+0xe4): undefined reference to `eCAL::experimental::measurement::hdf5::Reader::Reader(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```
Is a separate linking necessary for this? I currently use the following only in my CMakeLists for eCAL:

```cmake
find_package(eCAL REQUIRED)
find_package(Protobuf REQUIRED)

...
...
...

target_link_libraries(${PROJECT_NAME} PRIVATE
  eCAL::core
  protobuf::libprotobuf
)

**Reply by DownerCase** [reply]
You need to link to `eCAL::measurement_hdf5` as well to use the experimental API
https://github.com/eclipse-ecal/ecal/blob/support/v5.13/contrib/measurement/hdf5/CMakeLists.txt

There is also the original [`eCAL::eh5::HDF5Meas`](https://github.com/eclipse-ecal/ecal/blob/support/v5.13/contrib/ecalhdf5/include/ecalhdf5/eh5_meas.h) which has similar functions. The CMake target for that is `eCAL::hdf5`

**Reply by DownerCase** [reply]
> The [sample you shared above](https://github.com/eclipse-ecal/ecal/blob/support/v5.13/samples/cpp/measurement/measurement_read/src/measurement_read.cpp) has the structure `IMeasurement` from the header file `<ecal/measurement/imeasurement.h>`. Is this one relevant?

I only linked that as its the only reference I could find that wasn't the source code itself.

**Reply by piyush-m0** [reply]
> `eCAL::measurement_hdf5`

I will try this.
